File: test/functional/test_framework/siphash.py
#!/usr/bin/env python3
# Copyright (c) 2016-2018 The Bitcoin Core developers
# Distributed under the MIT software license, see the accompanying
# file COPYING or http://www.opensource.org/licenses/mit-license.php.
"""Specialized SipHash-2-4 implementations.

This implements SipHash-2-4 for 256-bit integers.
"""

def rotl64(n, b):
    return n >> (64 - b) | (n & ((1 << (64 - b)) - 1)) << b

def siphash_round(v0, v1, v2, v3):
    v0 = (v0 + v1) & ((1 << 64) - 1)
    v1 = rotl64(v1, 13)
    v1 ^= v0
    v0 = rotl64(v0, 32)
    v2 = (v2 + v3) & ((1 << 64) - 1)
    v3 = rotl64(v3, 16)
    v3 ^= v2
    v0 = (v0 + v3) & ((1 << 64) - 1)
    v3 = rotl64(v3, 21)
    v3 ^= v0
    v2 = (v2 + v1) & ((1 << 64) - 1)
    v1 = rotl64(v1, 17)
    v1 ^= v2
    v2 = rotl64(v2, 32)
    return (v0, v1, v2, v3)

def siphash256(k0, k1, h):
    n0 = h & ((1 << 64) - 1)
    n1 = (h >> 64) & ((1 << 64) - 1)
    n2 = (h >> 128) & ((1 << 64) - 1)
    n3 = (h >> 192) & ((1 << 64) - 1)
    v0 = 0x736f6d6570736575 ^ k0
    v1 = 0x646f72616e646f6d ^ k1
    v2 = 0x6c7967656e657261 ^ k0
    v3 = 0x7465646279746573 ^ k1 ^ n0
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    v0 ^= n0
    v3 ^= n1
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    v0 ^= n1
    v3 ^= n2
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    v0 ^= n2
    v3 ^= n3
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    v0 ^= n3
    v3 ^= 0x2000000000000000
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    v0 ^= 0x2000000000000000
    v2 ^= 0xFF
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    v0, v1, v2, v3 = siphash_round(v0, v1, v2, v3)
    return v0 ^ v1 ^ v2 ^ v3

File: src/crypto/siphash.cpp
// Copyright (c) 2016-present The Bitcoin Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include <crypto/siphash.h>

#include <uint256.h>

#include <cassert>
#include <span>

CSipHasher::CSipHasher(uint64_t k0, uint64_t k1) : m_state{k0, k1} {}

CSipHasher& CSipHasher::Write(uint64_t data)
{
    assert(m_count % 8 == 0);
    m_state.Compress2(data);
    m_count += 8;
    return *this;
}

CSipHasher& CSipHasher::Write(std::span<const unsigned char> data)
{
    SipHashState state{m_state.Copy()};
    uint64_t t{m_tmp};
    uint8_t c{m_count};

    while (data.size() > 0) {
        t |= uint64_t{data.front()} << (8 * (c % 8));
        c++;
        if ((c & 7) == 0) {
            state.Compress2(t);
            t = 0;
        }
        data = data.subspan(1);
    }

    m_state = state;
    m_count = c;
    m_tmp = t;

    return *this;
}

uint64_t CSipHasher::Finalize() const
{
    return m_state.Copy()
                  .Compress2(m_tmp | (uint64_t{m_count} << 56))
                  .Finalize4();
}

SipHasher13UJ& SipHasher13UJ::Write(uint64_t data) noexcept
{
    m_state.Compress1(data);
    return *this;
}

SipHasher13UJ& SipHasher13UJ::WriteJumbo(const uint256& hash) noexcept
{
    m_state.Compress1Jumbo(hash);
    return *this;
}

uint64_t SipHasher13UJ::Finalize() const noexcept
{
    return m_state.Copy().Finalize3U();
}

uint64_t PresaltedSipHasher::operator()(const uint256& val) const noexcept
{
    return m_state.Copy()
                  .Compress2(val.GetUint64(0))
                  .Compress2(val.GetUint64(1))
                  .Compress2(val.GetUint64(2))
                  .Compress2(val.GetUint64(3))
                  .Compress2(uint64_t{32} << 56)
                  .Finalize4();
}

uint64_t PresaltedSipHasher::operator()(const uint256& val, uint32_t extra) const noexcept
{
    return m_state.Copy()
                  .Compress2(val.GetUint64(0))
                  .Compress2(val.GetUint64(1))
                  .Compress2(val.GetUint64(2))
                  .Compress2(val.GetUint64(3))
                  .Compress2((uint64_t{36} << 56) | extra)
                  .Finalize4();
}
